control-plane-pki-operator/certificaterevocationcontroller/certificaterevocationcontroller.go

```diff
@@ -0,0 +1,683 @@
+		// a revocation timestamp exists, so we need to ensure new certs are generated after that point
+		signer, ok := secretForSignerClass(namespace, certificates.SignerClass(crr.Spec.SignerClass))
+		if !ok {
+			// we should never reach this case as we validate the class before transitioning states, and it's immutable
+			return nil, false, nil
+		}
+
+		certs, err := c.loadCertificateSecret(signer.Namespace, signer.Name)
+		if err != nil {
+			return nil, false, err
+		}
+
+		var certificateNeedsRegeneration bool
+		if len(certs) > 0 {
+			onlyAfter, beforeAndDuring := partitionCertificatesByValidity(certs, crr.Status.RevocationTimestamp.Time)
+			certificateNeedsRegeneration = len(beforeAndDuring) > 0 || len(onlyAfter) == 0
+		} else {
+			// there's no certificate there, regenerate it
+			certificateNeedsRegeneration = true
+		}
+
+		if certificateNeedsRegeneration {
+			// when we revoke a signer, we need to keep a copy of a previous leaf to verify that it's
+			// invalid when we're done revoking it
+			current, err := c.getSecret(signer.Namespace, signer.Name)
+			if err != nil {
+				return nil, false, err
+			}
+
+			// base36(sha224(value)) produces a useful, deterministic value that fits the requirements to be
+			// a Kubernetes object name (honoring length requirement, is a valid DNS subdomain, etc)
+			hash := sha256.Sum224([]byte(crr.Name))
+			var i big.Int
+			i.SetBytes(hash[:])
+			previousSignerName := i.Text(36)
+			copied, err := c.getSecret(namespace, previousSignerName)
+			if err != nil && !apierrors.IsNotFound(err) {
+				return nil, false, err
+			}
+			if apierrors.IsNotFound(err) ||
+				!reflect.DeepEqual(copied.Data[corev1.TLSCertKey], current.Data[corev1.TLSCertKey]) ||
+				!reflect.DeepEqual(copied.Data[corev1.TLSPrivateKeyKey], current.Data[corev1.TLSPrivateKeyKey]) {
+				secretCfg := corev1applyconfigurations.Secret(previousSignerName, signer.Namespace).
+					WithOwnerReferences(metav1applyconfigurations.OwnerReference().
+						WithAPIVersion(hypershiftv1beta1.GroupVersion.String()).
+						WithKind("CertificateRevocationRequest").
+						WithName(crr.Name).
+						WithUID(crr.UID)).
+					WithType(corev1.SecretTypeTLS).
+					WithData(map[string][]byte{
+						corev1.TLSCertKey:       current.Data[corev1.TLSCertKey],
+						corev1.TLSPrivateKeyKey: current.Data[corev1.TLSPrivateKeyKey],
+					})
+				return &actions{secret: secretCfg}, false, nil
+			}
+
+			if crr.Status.PreviousSigner == nil {
+				cfg := hypershiftv1beta1applyconfigurations.CertificateRevocationRequest(name, namespace)
+				cfg.Status = hypershiftv1beta1applyconfigurations.CertificateRevocationRequestStatus().
+					WithPreviousSigner(corev1.LocalObjectReference{Name: previousSignerName})
+				return &actions{crr: cfg}, false, nil
+			}
+
+			secretCfg := corev1applyconfigurations.Secret(signer.Name, signer.Namespace).
+				WithAnnotations(map[string]string{
+					certrotation.CertificateNotAfterAnnotation: "force-regeneration",
```

Comment:
are you planning on forcing serial handling of CRRs?

Reply:
Why would that be necessary?

Reply:
Didn't say it was necessary, asked if you were planning it.  It's easier to reason about serially, but perhaps if status is derivable from observing cluster state and spec it's easy enough to figure out.